src/extraction/FStar.Extraction.ML.Term.fst

```diff
@@ -671,33 +700,45 @@ let maybe_reify_term (env:TcEnv.env) (t:S.term) (l:lident) : S.term  =
   | S.Extract_none s ->
     err_cannot_extract_effect l t.pos s (Print.term_to_string t)
 
+let has_extract_as_impure_effect (g:uenv) (fv:S.fv) =
+  TcEnv.fv_has_attr (tcenv_of_uenv g) fv FStar.Parser.Const.extract_as_impure_effect_lid
+
+let head_of_type_is_extract_as_impure_effect g t =
+  let hd, _ = U.head_and_args t in
+  match (U.un_uinst hd).n with
+  | Tm_fvar fv -> has_extract_as_impure_effect g fv
+  | _ -> false
+
 let rec translate_term_to_mlty (g:uenv) (t0:term) : mlty =
     let arg_as_mlty (g:uenv) (a, _) : mlty =
         if is_type g a //This is just an optimization; we could in principle always emit MLTY_Erased, at the expense of more magics
         then translate_term_to_mlty g a
         else MLTY_Erased
     in
-
     let fv_app_as_mlty (g:uenv) (fv:fv) (args : args) : mlty =
         if not (is_fv_type g fv)
         then MLTY_Top //it was translated as an expression or erased
-        else
-            let formals, _ =
+        else (
+            if has_extract_as_impure_effect g fv
+            then let (a, _)::_ = args in
+                 translate_term_to_mlty g a
```

Comment:
Just extract the first argument (the result type) in case fv has the "extract_as_impure_effect" attribute